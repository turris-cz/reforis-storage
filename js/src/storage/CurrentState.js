/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import {
    Alert, ALERT_TYPES, RebootButton, SpinnerElement,
} from "foris";

import { NOT_PENDING_STORAGE_STATES, STORAGE_STATES } from "./constants";

export default function CurrentState({
    state, raid, uuid, old_uuid, old_device_desc, storageIsPending,
}) {
    const pendingMigration = old_uuid !== uuid;
    const brokenSetup = old_uuid === "broken";
    return (
        <>
            {!brokenSetup && pendingMigration
            && <PendingMigration old_uuid={old_uuid} old_device_desc={old_device_desc} />}

            {brokenSetup
            && <BrokenSetup />}

            <table className="table table-borderless table-hover offset-lg-1 col-lg-10 col-sm-12">
                <tbody>
                    {!Object.keys(NOT_PENDING_STORAGE_STATES).includes(state) && (
                        <tr>
                            <th scope="row">{_("State")}</th>
                            <td style={{ display: "flex", flexDirection: "row" }}>
                                {storageIsPending ? (
                                    <SpinnerElement small>
                                        {STORAGE_STATES[state]}
                                    </SpinnerElement>
                                ) : STORAGE_STATES[state]}
                            </td>
                        </tr>
                    )}
                    {old_device_desc !== "none"
                    && (
                        <tr>
                            <th scope="row">{_("Device")}</th>
                            <td>{old_device_desc}</td>
                        </tr>
                    )}
                    <tr>
                        <th scope="row">{_("UUID")}</th>
                        <td>{uuid}</td>
                    </tr>
                    <tr>
                        <th scope="row">{_("RAID")}</th>
                        <td>{raid}</td>
                    </tr>
                </tbody>
            </table>
        </>
    );
}

const PENDING_MIGRATION_TEXT = _(`
Your router is <b>waiting for restart</b> to proceed with migrating your data to the newly selected drive. Please
restart your router <b>as soon as possible</b> but bear in mind that the next reboot will take longer as all data need
to be migrated.
`);

function PendingMigration() {
    return (
        <Alert type={ALERT_TYPES.WARNING}>
            <p dangerouslySetInnerHTML={{ __html: PENDING_MIGRATION_TEXT }} />
            <RebootButton forisFormSize />
        </Alert>
    );
}

function BrokenSetup() {
    return (
        <Alert>
            <span>
                {_("Your setup is currently broken and you are probably loosing data, set a new storage device as soon as you can!")}
            </span>
        </Alert>
    );
}
