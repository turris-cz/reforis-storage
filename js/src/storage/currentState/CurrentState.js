/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import PropTypes from "prop-types";

import PendingMigrationAlert from "./PendingMigrationAlert";
import BrokenSetupAlert from "./BrokenSetupAlert";
import CurrentStateTable from "./CurrentStateTable";

CurrentState.propTypes = {
    state: PropTypes.string.isRequired,
    raid: PropTypes.string.isRequired,
    uuid: PropTypes.string.isRequired,
    old_uuid: PropTypes.string.isRequired,
    old_device_desc: PropTypes.string.isRequired,
    storageIsPending: PropTypes.bool.isRequired,
};

CurrentState.defaultProps = {
    storageIsPending: false,
};

export default function CurrentState({
    state, raid, uuid, old_uuid, old_device_desc, storageIsPending,
}) {
    const pendingMigration = old_uuid !== uuid;
    const brokenSetup = old_uuid === "broken";
    return (
        <>
            {!brokenSetup && pendingMigration && <PendingMigrationAlert />}

            {brokenSetup && <BrokenSetupAlert />}

            <CurrentStateTable
                state={state}
                old_device_desc={old_device_desc}
                uuid={uuid}
                raid={raid}
                storageIsPending={storageIsPending}
            />
        </>
    );
}
