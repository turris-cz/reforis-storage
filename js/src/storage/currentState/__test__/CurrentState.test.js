/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, getByText } from "foris/testUtils/customTestRender";

import CurrentState from "../CurrentState";
import getStateFixture from "../../__tests__/__fixtures__/getStateFixture";

describe("<CurrentState />", () => {
    let container;
    let rerender;

    beforeEach(() => {
        ({ container, rerender } = render(
            <CurrentState
                storageIsPending={false}
                {...getStateFixture()}
                disk_mounted
            />
        ));
    });

    it("Should render.", () => {
        expect(container).toMatchSnapshot();
    });

    it("Should render <PendingMigration/> on differnet uuid and old_uuid.", () => {
        rerender(
            <CurrentState
                storageIsPending={false}
                {...getStateFixture()}
                uuid="111"
                old_uuid="222"
                disk_mounted
            />
        );
        getByText(container, /waiting for restart/);
    });

    it("Should render <BrokenSetupAlert/> old_uuid broken.", () => {
        rerender(
            <CurrentState
                storageIsPending={false}
                {...getStateFixture()}
                old_uuid="broken"
                disk_mounted
            />
        );
        getByText(container, /Your setup is currently broken/);
    });
});
