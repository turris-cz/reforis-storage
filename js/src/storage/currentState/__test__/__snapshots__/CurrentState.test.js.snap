// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CurrentState /> should render 1`] = `
<div>
  <table
    class="table table-borderless table-hover offset-lg-1 col-lg-10 col-sm-12"
  >
    <tbody>
      <tr>
        <th
          scope="row"
        >
          Device
        </th>
        <td>
          /dev/sdb
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          UUID
        </th>
        <td>
          1234-1234-1234-1234-1234
        </td>
      </tr>
      <tr>
        <th
          scope="row"
        >
          RAID
        </th>
        <td>
          single
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
